firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/crcelements.txt

```diff
@@ -0,0 +1,131 @@
+      THEN("The output buffer is as expected")
+
+/* parse function
+Scenario: CRCElement: The parse method correctly updates its internal crc value and the constructor's payload buffer from the input buffer
+  GIVEN("A CRCElement constructed with an empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body without a complete 4-byte header is parsed")
+      THEN("the parse method reports out of bounds status")
+
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is parsed")
+      THEN("the parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor") 
+
+    WHEN("A body with a complete 4-byte header, and a payload inconsistent with the CRC field of the header, is parsed")
+      THEN("the parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the value returned by the crc accessor method is inconsistent with the payload")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+
+  GIVEN("A CRCElement constructed with a non-empty payload buffer of capacity 254 bytes")
+    WHEN("A body with a complete 4-byte header, and a payload consistent with the CRC field of the header, is parsed")
+      THEN("the crc accessor method returns 0 before the parse method is called")
+      THEN("the payload returned from the payload accessor method is equal to the payload buffer given in the CRCElement constructor")
+      THEN("the parse method reports ok status")
+      THEN("the value returned by the crc accessor method is equal to the crc field of the input body's header")
+      THEN("the payload returned from the payload accessor method is equal to the payload from the body")
+      THEN("the payload given in the CRCElement constructor is independent of the input body")
+      THEN("the payload accessor method returns a reference to the payload buffer given in the CRCElement constructor")
+
+/* write parse roundtrip
+Scenario: CRCElement correctly preserves data in roundtrip writing and parsing
+
+  GIVEN("A CRC Element initialized with a non empty payload buffer with a capacity of 254 bytes")
+
+    WHEN("A body with crc and payload is generated, parsed and generated again")
```

Comment:
I thought about this a bit more and I think the test would be stronger if we didn't reuse the same CRCElement object (and the same payload buffer) for all three write/parse/write method calls, but instead instantiated a new ParsedCRCElement/ConstructedCRCElement object (and a new payload buffer, for ParsedCRCElement objects) for each method call. This way we can ensure that roundtripping works correctly even if no internal state is preserved across those method calls, which is closer to how the objects would be instantiated and used in the application.

Then we could update the WHEN as: WHEN A body with CRC and payload is generated, parsed into a new ParsedCRCElement object, and generated again with a new ConstructedCRCElement object

Reply:
Updated using ParsedCRCElement & ConstructedCRCElement, using different payload buffers for constructing the elements and different buffers for input.